### ISSUE TITLE
Please implement multiselect and "Copy with Headers"

### ISSUE THREAD

**Issue by dburtsev** [JAM-Software/Virtual-TreeView#1266]
Hello
Can you please implement cells multi-select and "Copy with Headers" functionality?
![Screenshot 2024-09-12 092513](https://github.com/user-attachments/assets/20447f4b-0b1c-4dc1-a4f1-22918b682dba)

**Comment by chuacw** [comment]
How does Copy with Headers and Copy without Headers work? What format would they be in?

**Comment by dburtsev** [comment]
How can I check format? I can copy and paste into Notepad++ and Excel.

[CopyHearer.docx](https://github.com/user-attachments/files/24454245/CopyHearer.docx)

**Comment by chuacw** [comment]
Use http://www.peterbuettner.de/develop/tools/clipview/ and check all available formats, as well as all raw data that you want supported, and provide them here.

**Comment by dburtsev** [comment]
Done.

[Doc1.docx](https://github.com/user-attachments/files/24455020/Doc1.docx)

**Comment by chuacw** [comment]
@dburtsev 

Please provide data directly in the comments, not in docx, as I cannot view docx.
Is the original data that you were copying in the screenshot in the first comment? If not, please provide the data that you were copying. 

Ensure all the data can be seen.

Thank you.

**Comment by dburtsev** [comment]
Main SQL query: SELECT table_schema, table_name, table_type FROM information_schema.tables;
49319: HTML Format
Version:1.0
StartHTML:0000000105
EndHTML:0000000551
StartFragment:0000000201
EndFragment:0000000535
<html><head><meta http-equiv="Content-Type" content="text/html;charset=utf-8" ></head>
<body>

<table><tr><th>table_name</th><th>table_type</th></tr><tr><td>pg_settings</td><td>VIEW</td></tr><tr><td>pg_type</td><td>BASE TABLE</td></tr><tr><td>pg_attribute</td><td>BASE TABLE</td></tr><tr><td>pg_class</td><td>BASE TABLE</td></tr><tr><td>pg_tablespace</td><td>BASE TABLE</td></tr></table>

</body></html>

49954: System.String
–§žý;pC¦yVk²ˆû

13: CF_UNICODETEXT
table_name	table_type
pg_settings	VIEW
pg_type	BASE TABLE
pg_attribute	BASE TABLE
pg_class	BASE TABLE
pg_tablespace	BASE TABLE

1: CF_TEXT
table_name	table_type
pg_settings	VIEW
pg_type	BASE TABLE
pg_attribute	BASE TABLE
pg_class	BASE TABLE
pg_tablespace	BASE TABLE

7: CF_OEMTEXT
table_name	table_type
pg_settings	VIEW
pg_type	BASE TABLE
pg_attribute	BASE TABLE
pg_class	BASE TABLE
pg_tablespace	BASE TABLE

<img width="599" height="343" alt="Image" src="https://github.com/user-attachments/assets/7c8ba8b2-4bf8-4f3d-acac-d1872d3c8033" />

**Comment by chuacw** [comment]
Please test this branch for multicell selection (no copy for now)
https://github.com/chuacw/Virtual-TreeView/tree/Multiselect

Feedback in that branch. Once stable, I'll send a PR here.

**Comment by dburtsev** [comment]
Sorry I don't have Delphi.

**Comment by chuacw** [comment]
I'm confused. How are you using these components then?

**Comment by dburtsev** [comment]
I am using HeidiSQL which is using this grid. https://github.com/HeidiSQL/HeidiSQL

**Comment by chuacw** [comment]
In that case, you should file a request with HeidiSQL. Whatever is implemented here, there's no guarantee they'll use it.

**Comment by joachimmarder** [comment]
> Please test this branch for multicell selection (no copy for now)

Thank you very much, @chuacw. So far I did not have time to review the chnages. Am I assuming correct that this feature will be enabled / used only in case the the option `TVTMiscOption.toGridExtensions` *and* `TVTSelectionOption.toMultiSelect` is enabled, and `TVTSelectionOption.toFullRowSelect` is not enabled? That would make sure the new code gets only involved in specific situations. It coould make sense to extract this decision in a small virtual helper function `IsCellSelectionEnabled()` as a single point of truth, where one can easily ovverride 

And I would like to suggest that I make a release of [V8.3](https://github.com/JAM-Software/Virtual-TreeView/milestone/33?closed=1) soon, so that we have a stable release and then merge your PR for V8.4.

**Comment by chuacw** [comment]
Hi @joachimmarder ,

You're half correct.

The multicell selection will only be enabled when toMultiSelect, toExtendedFocus is included.
In addition, toFullRowSelect must be excluded from TreeOptions.SelectionOptions.
If you refer to  [this line](https://github.com/chuacw/Virtual-TreeView/blob/14b60a0bfc8d3bbf0cdc57a616b84fc57d6451e1/Source/VirtualTrees.BaseTree.pas#L1122), it'll show that.

This is because, by definition, the toFullRowSelect option doesn't make sense when selecting multiple cells.

I've marked various parts of the code with multicell so you can review it easily.

Thanks for the feedback on the small virtual helper function. I'll add that.

There's both a demo: Multiselect.dpr that demonstrates multicell selection and copy into the clipboard, as well as a new unit test.

Sure, no problem on merging into 8.4.

**Comment by joachimmarder** [comment]
Hi @chuacw,

I tested your demo project and it looks good. I only missed the ability to use Shift + Cursor to perform a cell selection, like in other grids or Excel. With that feature included I would be happy to merge your fork in the master branch. Thank you!

**Comment by chuacw** [comment]
Hi @joachimmarder 

Can you clarify? By shift + cursor, do you mean shift + arrow keys?

**Comment by chuacw** [comment]
The latest commit fixed selection with shift + arrow keys.

**Comment by joachimmarder** [comment]
> Can you clarify? By shift + cursor, do you mean shift + arrow keys?

Yes, arrow key, sorry, im German you often refer to these as "curor keys" (see [Wikipedia](https://de.wikipedia.org/wiki/Pfeiltaste#:~:text=Als%20Pfeiltasten%2C%20auch%20Cursortasten%2C%20werden,nahezu%20allen%20tastenorientierten%20Eingabeger%C3%A4ten.)). 

Thanks for adding this support!

**Comment by dburtsev** [comment]
@chuacw
The author of HeidiSQL points me to this repository. 
He will enable this functionality in the future version of HeidiSQL if you implement it in the grid.
Thank you

**Comment by joachimmarder** [comment]
@chuacw : Your PR is now merged - thank you!
